adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -725,43 +1150,124 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
-    genomicRdd: GenomicRDD[X, Y],
-    optPartitions: Option[Int] = None)(
-      implicit tTag: ClassTag[T], xTag: ClassTag[X]): GenomicRDD[(Option[T], Iterable[X]), Z] = {
+  def rightOuterShuffleRegionJoinAndGroupByLeft[X, Y <: GenomicRDD[X, Y], Z <: GenomicRDD[(Option[T], Iterable[X]), Z]](genomicRdd: GenomicRDD[X, Y],
+                                                                                                                        optPartitions: Option[Int] = None)(
+                                                                                                                          implicit tTag: ClassTag[T], xTag: ClassTag[X]): GenomicRDD[(Option[T], Iterable[X]), Z] = {
+    // did the user provide a set partition count?
+    // if no, take the max partition count from our rdds
+    val partitions = optPartitions.getOrElse(Seq(rdd.partitions.length,
+      genomicRdd.rdd.partitions.length).max)
 
-    val (partitionSize, endSequences) = joinPartitionSizeAndSequences(optPartitions,
-      genomicRdd)
+    val leftRdd = repartitionAndSortByGenomicCoordinate(partitions)
+    leftRdd.SortedTrait.rightOuterShuffleRegionJoinAndGroupByLeft(genomicRdd, optPartitions)
+  }
 
-    // key the RDDs and join
-    GenericGenomicRDD[(Option[T], Iterable[X])](
-      RightOuterShuffleRegionJoinAndGroupByLeft[T, X](endSequences,
-        partitionSize,
-        rdd.context).partitionAndJoin(flattenRddByRegions(),
-          genomicRdd.flattenRddByRegions()),
-      endSequences,
-      kv => {
-        (kv._2.flatMap(v => genomicRdd.getReferenceRegions(v)) ++
-          kv._1.toSeq.flatMap(v => getReferenceRegions(v))).toSeq
-      })
-      .asInstanceOf[GenomicRDD[(Option[T], Iterable[X]), Z]]
+  /**
+   * The purpose of this method is to co-partition two RDDs according to their ReferenceRegions.
+   * It is useful during ShuffleRegionJoins because it guarantees no extra shuffling to get correct
+   * and complete region joins. RDD is already sorted coming in, so we take precaution to keep it
+   * that way.
+   *
+   * The simplest way of sending data to the correct node is to just send everything that has a
+   * start position that could fall within the boundaries
+   *
+   * This is best used under the condition that (repeatedly) repartitioning is more expensive than
+   * calculating the proper location of the records of this.rdd. It requires a pass through the
+   * co-located RDD to get the correct partition(s) for each record. It will assign a record to
+   * multiple partitions if necessary.
+   *
+   * @param rddToCoPartitionWith the rdd that this.rdd will be copartitioned with
+   * @return returns the newly repartitioned rdd that is co-partitioned with rddToCoPartitionWith
+   */
+  def coPartitionByGenomicRegion[X, Y <: GenomicRDD[X, Y]](rddToCoPartitionWith: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): U = {
```

Comment:
"copartition" is a single word, `coPartitionByGenomicRegion` &rarr; `copartitionByGenomicRegion`